[CLI] Shorthand notation to instantiate callbacks [3/3] (#8815)
PyTorchLightning/pytorch-lightning: docs/source/common/lightning_cli.rst

:class:`~pytorch_lightning.core.datamodule.LightningDataModule` classes that have as type hint a class can be configured
the same way using :code:`class_path` and :code:`init_args`.

> Lightning optionally simplifies the user command line so that only the :class:`~pytorch_lightning.callbacks.Callback` name is required.

Reviewer: Add a subtitle here, maybe "Register your callbacks" or something like that

The argument's order matters and the user needs to pass the arguments in the following way.
This is supported only for PyTorch Lightning built-in :class:`~pytorch_lightning.callbacks.Callback`.

.. code-block:: bash

    $  python ... --trainer.callbacks={CALLBACK_NAME_1} --trainer.callbacks.{CALLBACK_1_ARGS_1}=... --trainer.callbacks.{CALLBACK_1_ARGS_2}=... --trainer.callbacks={CALLBACK_N} --trainer.callbacks.{CALLBACK_N_ARGS_1}=...

Here is an example:

.. code-block:: bash

    $  python ... --trainer.callbacks=EarlyStopping --trainer.callbacks.patience=5 --trainer.callbacks=LearningRateMonitor

> However, a user can register their own callbacks as follows.

Reviewer: ```suggestion
To register your own callback:
```

Author: Done !

:class:`~pytorch_lightning.core.datamodule.LightningDataModule` classes that have as type hint a class can be configured
the same way using :code:`class_path` and :code:`init_args`.

Alternatively, the user can provide the list of callbacks directly from the command line.

.. code-block:: bash

    $  python ... --trainer.callbacks=[{"class_path": pytorch_lightning.callbacks.EarlyStopping, "init_args": "patience": "5"}, ...].

> Lightning optionally simplifies the user command line where only the :class:`~pytorch_lightning.callbacks.Callback` name is required.

Reviewer 2: ```suggestion
Lightning optionally simplifies the user command line so that only the :class:`~pytorch_lightning.callbacks.Callback` name is required.
```

Author: Done !

The argument's order matters and the user needs to pass the arguments in the following way.
This is supported only for PyTorch Lightning built-in :class:`~pytorch_lightning.callbacks.Callback`.

.. code-block:: bash

>     $  python ... --trainer.callbacks={CALLBACK_NAME_1} --trainer.{CALLBACK_1_ARGS_1}=... --trainer.{CALLBACK_1_ARGS_2}=... --trainer.callbacks={CALLBACK_N} --trainer.{CALLBACK_N_ARGS_1}=...

Author: Oh this was a doc typo.

Should be:

```bash
--trainer.callbacks={CALLBACK_NAME_1} --trainer.callbacks.{CALLBACK_1_ARGS_1}=...
```

Here is an example:

.. code-block:: bash

>     $  python ... --trainer.callbacks=EarlyStopping --trainer.patience=5 --trainer.callbacks=LearningRateMonitor

Author: @carmocca Sorry, there were a typo in the docs.

Should have been:
```
$  python ... --trainer.callbacks=EarlyStopping --trainer.callbacks.patience=5 --trainer.callbacks=LearningRateMonitor
```

I thought about `--trainer.EarlyStopping.patience=5`, but I really don't like the design. I know some users will complain about it.

> However, a user can register its own callbacks as follow.

Reviewer 2: ```suggestion
However, a user can register their own callbacks as follows.
```

Author: Done !

.. code-block:: python

    from pytorch_lightning.utilities.cli import CALLBACK_REGISTRY
    from pytorch_lightning.callbacks import Callback

    @CALLBACK_REGISTRY
    class CustomCallback(Callback):

>         pass

Author: Done !

    $ python train.py --optimizer.class_path=torch.optim.Adam --optimizer.init_args.lr=0.01

> Optionally, the command line can be simplified for PyTorch built-in `optimizers` and `schedulers`:

Reviewer: It sounds to me like this should start the "Optimizers and learning rate schedulers" section. Also, let's add subtitles here to make it easier to find

Built in schedulers & optimizers
-------------------------------
If You need pytorch built in optimizers or scehdulers -> do this.

Register your own bla bla
-------------------------

Configure optimizers from CLI
------------------------------

    cli = MyLightningCLI(MyModel)

For both possibilities of using :meth:`pytorch_lightning.utilities.cli.LightningArgumentParser.add_optimizer_args` with
a single class or a tuple of classes, the value given to :code:`optimizer_init` will always be a dictionary including
:code:`class_path` and :code:`init_args` entries. The function
:func:`~pytorch_lightning.utilities.cli.instantiate_class` takes care of importing the class defined in
:code:`class_path` and instantiating it using some positional arguments, in this case :code:`self.parameters()`, and the
:code:`init_args`. Any number of optimizers and learning rate schedulers can be added when using :code:`link_to`.

Built in schedulers & optimizers and registering your own
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For code simplification, the LightningCLI provides properties with PyTorch's built-in `optimizers` and `schedulers` already registered.

Only the optimizer or scheduler name needs to be passed along its arguments.

.. code-block:: bash

    $ python train.py --optimizer=Adam --optimizer.lr=0.01 --lr_scheduler=CosineAnnealingLR

> If you model requires multiple optimizers, the LightningCLI provides already registered optimizers and schedulers under the properties `registered_optimizers` and `registered_lr_schedulers`

Author: Done !

    cli = MyLightningCLI(MyModel)

For both possibilities of using :meth:`pytorch_lightning.utilities.cli.LightningArgumentParser.add_optimizer_args` with
a single class or a tuple of classes, the value given to :code:`optimizer_init` will always be a dictionary including
:code:`class_path` and :code:`init_args` entries. The function
:func:`~pytorch_lightning.utilities.cli.instantiate_class` takes care of importing the class defined in
:code:`class_path` and instantiating it using some positional arguments, in this case :code:`self.parameters()`, and the
:code:`init_args`. Any number of optimizers and learning rate schedulers can be added when using :code:`link_to`.

> For code simplification, the LightningCLI provides properties with already registered PyTorch built-in `optimizers` and `schedulers`.

Reviewer 2: ```suggestion
For code simplification, the LightningCLI provides properties with PyTorch's built-in `optimizers` and `schedulers` already registered.
```

Author: Done !

.. code-block::

    class MyLightningCLI(LightningCLI):
        def add_arguments_to_parser(self, parser):
            parser.add_optimizer_args(
                self.registered_optimizers,
                link_to="model.optimizer_init",
            )
            parser.add_lr_scheduler_args(
                self.registered_lr_schedulers,
                link_to="model.lr_scheduler_init",
            )

> However, a user can register its own optimizers or schedulers as follows.

Reviewer 2: ```suggestion
However, a user can register their own optimizers or schedulers as follows.
```

Author: Done !

.. code-block:: python

    import torch
    from pytorch_lightning.utilities.cli import OPTIMIZER_REGISTRIES, LR_SCHEDULER_REGISTRIES
    from pytorch_lightning.callbacks import Callback

>     @CALLBACK_REGISTRIES

Reviewer 3: ```suggestion
    @OPTIMIZER_REGISTRIES
```

Reviewer 4: Hopefully this becomes `@OPTIMIZER_REGISTRY` right?

Reviewer 3: Yup

Author: Done !
